jupyterhub_chameleon/handler.py

```diff
@@ -184,76 +137,116 @@ async def _fetch_new_token(self, refresh_token):
+            trovi_token = trovi.exchange_token(auth_state.get("access_token"))
 
-        auth_state = await self.current_user.get_auth_state()
+            artifact = trovi.fetch_artifact(artifact_id, trovi_token)
 
-        # Authenticate to Trovi with our OpenStack credentials,
-        # exchanging our keystone token for a Trovi token
-        trovi_resp = requests.post(
-            urljoin(os.getenv("TROVI_URL"), "/token/"),
-            headers={"Content-Type": "application/json", "Accept": "application/json"},
-            json={
-                "grant_type": "token_exchange",
-                "subject_token": auth_state.get("access_token"),
-                "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
-                "scope": "artifacts:read artifacts:write",
-            },
-        )
+            versions = list(
+                sorted(
+                    artifact["versions"],
+                    reverse=True,
+                    key=lambda v: v["created_at"],
+                )
+            )
+            if version_number < 0 or version_number >= len(versions):
+                raise HTTPError(requests.codes.bad_request, "Invalid version number")
+            if len(versions) == 0:
+                raise HTTPError(
+                    requests.codes.internal_server_error,
+                    "Artifact has no linked content.",
+                )
+            version = versions[version_number]
 
-        trovi_token = trovi_resp.json()
+            contents_url = trovi.fetch_contents_url(
+                version["contents"]["urn"], trovi_token
+            )
 
-        self.log.debug(auth_state.get("access_token"))
+            owner = artifact["owner_urn"].split(":")[-1]
+            if self.current_user.name == owner or not artifact.id:
+                self.ownership = "own"
+            else:
+                self.ownership = "fork"
 
-        self.log.debug(f"Trovi token exchange response: {trovi_token}")
+            self.artifact = artifact
+            self.version = version
+            self.contents_url = contents_url
 
-        if trovi_resp.status_code in (
-            requests.codes.unauthorized,
-            requests.codes.forbidden,
-        ):
-            self.log.error(f"Authentication to trovi failed: {trovi_token}")
-            raise HTTPError(
-                requests.codes.unauthorized,
-                "You are not authorized to upload artifacts to Trovi via Jupyter.",
-            )
-        elif trovi_resp.status_code != requests.codes.created:
-            self.log.error(f"Authentication to trovi failed: {trovi_token}")
-            raise HTTPError(
-                requests.codes.internal_server_error,
-                "Unknown error uploading artifact to Trovi.",
-            )
+            sha = hashlib.sha256()
+            sha.update(version["contents"]["urn"].encode("utf-8"))
+            server_name = sha.hexdigest()[:7]
```

Comment:
I'm not sure this sha thing is that useful anymore, maybe it would be better to have it be like `{artifact.id}-{version.id}`? or just version UUID or something. I forget if there is a length limit on the server names. The SHA is a bit annoying b/c as admins it can be hard to tell what artifact it's linked to just by name.